ext/liquid_c/vm.c

```diff
@@ -0,0 +1,75 @@
+#include <stdint.h>
+
+#include "liquid.h"
+#include "vm.h"
+
+ID id_render_node;
+ID id_ivar_interrupts;
+ID id_ivar_resource_limits;
+ID id_ivar_render_length_limit;
+ID id_ivar_last_capture_length;
+ID id_increment_render_score;
+ID id_increment_write_score;
+ID id_raise_limits_reached;
+
+void liquid_vm_render(block_body_t *body, VALUE context, VALUE output)
+{
+    Check_Type(output, T_STRING);
+
+    VALUE resource_limits = rb_ivar_get(context, id_ivar_resource_limits);
+    rb_funcall(resource_limits, id_increment_render_score, 1, INT2NUM(body->render_score));
+
+    bool is_captured = rb_ivar_get(resource_limits, id_ivar_last_capture_length) != Qnil;;
+    long render_length_limit = LONG_MAX;
+
+    if (!is_captured) {
+        VALUE render_length_limit_num = rb_ivar_get(resource_limits, id_ivar_render_length_limit);
+        if (render_length_limit_num != Qnil)
+            render_length_limit = NUM2LONG(render_length_limit_num);
+    }
+
+    const size_t *const_ptr = (const size_t *)body->constants.data;
+    const uint8_t *ip = body->instructions.data;
+    VALUE interrupts = rb_ivar_get(context, id_ivar_interrupts);
+    Check_Type(interrupts, T_ARRAY);
+
+    while (true) {
+        switch (*ip++) {
+            case OP_LEAVE:
+                return;
+            case OP_WRITE_RAW:
+            {
+                const char *text = (const char *)*const_ptr++;
```

Comment:
Very nice and simple way to avoid index overflow 👍 

However, it will break once we introduce loops (`for`, etc.)

Reply:
I do plan on revisiting the moving constant pointer when adding jumps.  It could be implemented by have the jump target be a pair of instruction and constant pointer offsets, so there is a trade-off between jump overhead and constant lookup overhead.

Reply:
Sounds good. First time seeing this approach...

But when you'll add a pointer offset, I think it will be equivalent to the traditional `constants[i]`, and you'll end up needing multi-byte index too.